### ISSUE TITLE
MIDI Output - Frontend MIDI Output not working (Retroarch)

### ISSUE THREAD

**Issue by gamerk2** [schellingb/dosbox-pure#89]
Here's a bit of an odd one: I have a number of DOS games configured to use either MUNT (for MT-32) or Sound Canvas VA (for SC-55/88) MIDI output, via loopMIDI to create virtual MIDI ports that can be accessed by DOSBOX.  My old configuration using DOXBOX SVN (standalone) works; all titles output MIDI as expected.

I gave DOSBOX Pure a try.  For the most part, neither MT-32 or SC-55/88 titles are outputting any MIDI to the loopMIDI port.  The only title I have that seems to work correctly is Duke Nukem 3D; the other titles I have tried (Wing Command [MT-32], Doom2 [SC-88], and Kings Quest VI [General MIDI/SC-88]) result in NULL output.

While not my normal configuration, I downloaded the DOSBOX SVN core through Retroarch and confirmed the MIDI output works correctly for these devices.  That would seem to isolate the problem to DOSBOX Pure since Retroarch/DOSBOX SVN works correctly.

What I find odd is that Duke Nukem 3D works, yet all other titles I have tested so far do not.

EDIT

Forgot to mention, this is all on the latest nightly of Retroarch, latest DOSBOX Pure (0.9).  Win 10 x64.

**Comment by sweatypickle** [comment]
Is Retroarch configured to use loopMIDI as your MIDI device?

**Comment by gamerk2** [comment]
Yes.  And the setup is confirmed working on DOSBOX SVN through Retroarch.

Using DOSBOX Pure, aside from Duke Nukem 3D (which works) none of my other
games are outputting MIDI audio through loopMIDI, and I don't see loopMIDI
reporting it's seeing any data (EG: the "Total Data" and "Throughput / sec"
fields aren't changing).

**Comment by gamerk2** [comment]
I have some new information:

I just finished up some tests, and discovered entirely by accident that toggling Retroarch's Fullscreen mode immediately fixes the problem; I immediately get MIDI as expected.  It doesn't seem to matter if I go Fullscreen->Window or Window->Fullscreen, either seems to work.  As far as I can tell, Alt-Tab doesn't seem to affect anything, nor does toggling Windowed Fullscreen.  This seems repeatable, and was able to confirm Wing Commander, Kings Quest VI, and Ultimate Doom now output to their MIDI synthesizers (MUNT and Sound Canvas VA) as expected.  I assume other titles will as well.

No idea why toggling Fullscreen is affecting audio output, but at least it gives something to look at.  The only other thing I see is I do record some data being sent to my MIDI synthesizers (a few KB) when a game is initially loaded, then nothing until I do a Fullscreen toggle.  The only thing I can think of is something is getting hosed during init that is getting reset when changing Retroarch's Fullscreen setting. *shrug*

DOSBOX SVN is not affected by this.

**Comment by schellingb** [comment]
Just to make sure, do you have the option "`Audio > MIDI Output`" of DOSBox Pure set to "`Frontend MIDI driver`" and the option "`Sound: Enable libretro MIDI passthrough`" of DOSBox-svn set to "ON"?

**Comment by gamerk2** [comment]
Yes.

After doing a ton of testing, I noticed that even DOSBOX SVN was running into problems with the RetroArch Nightly build I was using.  I updated to the latest Nightly (03/11/21) and things *appear* to be working correctly.  I tested a number of MT-32 and SC-55 titles and everything appears to be getting routed where it needs to go as far as MIDI audio.  So I'm going to blame the Retroarch Nightly build I was using, since everything looks to be OK now.

That being said: It is worth noting that Frontend audio does *not* work in Retroarch 1.9.0 stable; when attempting to use Frontend audio you get "Frontend Audio device is not configured correctly", even though it is configured the same way that works in the latest Nightly builds.
